openedx/core/djangoapps/content_libraries/docs/decisions/0004-upstream-block.rst

```diff
@@ -0,0 +1,89 @@
+4. Persist the relationship between blocks and their upstream source
+####################################################################
+
+Status
+******
+
+Accepted
+
+Pending implementation in (TODO - link)
+
+Context
+*******
+
+* TODO - context
+
+Decision
+********
+
+* The following two fields will be added to every XBlock in CMS:
+
+  .. code-block:: python
+
+     upstream_block = String(
+         scope=Scope.settings,
+         help=(
+             "The usage key of a the block that served as this block's template, if one exists. "
+             "Generally, the upstream_block is within a Content Library."
+         ),
+         hidden=True,
+         default=None,
+         ...
+     )
+     upstream_block_version = Integer(
+         scope=Scope.settings,
+         help=(
+             "The upstream_block's version number, at the time this block was created from it. "
+             "If this version is older than the upstream_block's latest version, then CMS will "
+             "allow this block to fetch updated content from upstream_block."
+         ),
+         hidden=True,
+         default=None,
+         ...
+     )
+
+* The upstream-defined values of XBlock settings will be exposed in the XBlock API (TODO - details)
+
+* The upstream-defined values will be set when the XBlock is initialized by (TODO - figure this out)
+
+* The existing ``get_block_original_usage`` Modulestore methods will (TODO - figure this out)
+
+* New attributes will be serialized into the OLX of blocks with upstreams:
+
+  * The two new XBlock fields will serialize into OLX as attributes, like any settings-scoped field.
+  * The upstream defaults will use a new special syntax: ``upstream.$FIELD_NAME``.
+  * The period (``.``) was explicitly chosen to ensure that these special attributes cannot possibly conflict with any Python-defined XBlock fields.
+
+    .. code-block:: xml
+
+       <problem
+         display_name="A title that has been customized in the course"
```

Comment:
@kdmccormick, @bradenmacdonald: Something that came up in a recent discussion is how a `display_name` setting would look for an item that was copied into a course generally, items that are pulled in as a child of a `LibraryContentBlock` would have to have their `display_name` set to an empty string because it's the `LibraryContentBlock` that will be providing the `display_name` in those cases.

Reply:
@ormsbee breaking this down by static vs randomized reference:

* static reference: we're not using LibraryContentBlock any more, so it seems fine to use the upstream `display_name`, and allow the course author to override it to `""`, right?
  * Note to self: Need to make sure we handle `""` as a legitimate override, rather than accidentally falling back to the upstream default just because empty string is falsy.
* randomized reference: This is now just an ItemPool wrapping a bunch of static references. I could imagine that a course author would want to use the ItemPool's display_name rather than each of its children's display names. Some ideas for how to handle this:
  * A `show_item_display_names` boolean setting on ItemPool; when False, it renders its children without their titles.
  * A button on the ItemPool editor which, in one swoop, overrides all of its children's titles to `""`. This is nice because it's totally at the UI level; no special data changes needed.

Reply:
> * static reference: we're not using LibraryContentBlock any more, so it seems fine to use the upstream display_name, and allow the course author to override it to "", right?

By this, you mean that there's going to be an `upstream.display_name` that has the original library-given name, and then a `display_name=""` set on the block in the course, right? If so, 👍 

> randomized reference: This is now just an ItemPool wrapping a bunch of static references. I could imagine that a course author would want to use the ItemPool's display_name rather than each of its children's display names. Some ideas for how to handle this:
> * A `show_item_display_names` boolean setting on ItemPool; when False, it renders its children without their titles.
> * A button on the ItemPool editor which, in one swoop, overrides all of its children's titles to `""`. This is nice because it's totally at the UI level; no special data changes needed.

I think either of these is fine from the technical side, and it's mostly a matter of UX preferences. That being said, I suspect we'll end up with something more like the `show_item_display_names` entry, because those display names might be useful in other contexts, e.g. "rescore problem X for all students because there was a bug we fixed" or "how did people perform on problem Y"?